pkg/runctl/controller.go

```diff
@@ -343,111 +326,126 @@ func (c *Controller) syncHandler(key string) error {
 			// Return error that the pipeline stays in the queue and will be processed after switching back to normal mode.
 			return err
 		}
-		if err = c.changeState(pipelineRun, api.StatePreparing); err != nil {
+		if err = c.changeAndCommitStateAndMeter(pipelineRun, api.StatePreparing, metav1.Now()); err != nil {
 			return err
 		}
 		c.metrics.CountStart()
 	}
 
 	runManager := c.createRunManager(pipelineRun)
 
-	// the configuration should be loaded once per sync to avoid inconsistencies
-	// in case of concurrent configuration changes
-	pipelineRunsConfig, err := c.loadPipelineRunsConfig()
-	if err != nil {
-		if serrors.IsRecoverable(err) {
-			c.recorder.Event(pipelineRunAPIObj, corev1.EventTypeWarning, api.EventReasonLoadPipelineRunsConfigFailed, err.Error())
-			return err
-		}
-		pipelineRun.UpdateResult(api.ResultErrorInfra)
-		pipelineRun.StoreErrorAsMessage(err, "failed to load configuration for pipeline runs")
-		c.metrics.CountResult(pipelineRun.GetStatus().Result)
-		runManager.Cleanup(pipelineRun)
-		return c.finish(pipelineRun)
-	}
-
 	// Process pipeline run based on current state
 	switch state := pipelineRun.GetStatus().State; state {
 	case api.StatePreparing:
-		err = runManager.Start(pipelineRun, pipelineRunsConfig)
+		// the configuration should be loaded once per sync to avoid inconsistencies
+		// in case of concurrent configuration changes
+		pipelineRunsConfig, err := c.loadPipelineRunsConfig()
+		if err != nil {
+			return c.onGetRunError(pipelineRunAPIObj, pipelineRun, err, api.StateFinished, api.ResultErrorInfra, "failed to load configuration for pipeline runs")
+		}
+		namespace, auxNamespace, err := runManager.Start(pipelineRun, pipelineRunsConfig)
 		if err != nil {
 			c.recorder.Event(pipelineRunAPIObj, corev1.EventTypeWarning, api.EventReasonPreparingFailed, err.Error())
 			resultClass := serrors.GetClass(err)
 			// In case we have a result we can cleanup. Otherwise we retry in the next iteration.
 			if resultClass != api.ResultUndefined {
 				pipelineRun.UpdateMessage(err.Error())
-				pipelineRun.UpdateResult(resultClass)
-				if errClean := c.changeState(pipelineRun, api.StateCleaning); errClean != nil {
-					return errClean
-				}
 				pipelineRun.StoreErrorAsMessage(err, "preparing failed")
-				c.metrics.CountResult(pipelineRun.GetStatus().Result)
-				return nil
+				return c.updateStateAndResult(pipelineRun, api.StateCleaning, resultClass, metav1.Now())
 			}
 			return err
 		}
-		if err = c.changeState(pipelineRun, api.StateWaiting); err != nil {
+
+		pipelineRun.UpdateRunNamespace(namespace)
+		pipelineRun.UpdateAuxNamespace(auxNamespace)
+
+		if err = c.changeAndCommitStateAndMeter(pipelineRun, api.StateWaiting, metav1.Now()); err != nil {
```

Comment:
The only potential issues is that if we fail in this step, namespace and auz-namepspace will be un-managed in the cluster. But this issue already occured in previous implementation. It is fine to let as it is. Still, I would suggest for those external resource, when it is created, the owner should be kept in this resource. Or just keep special labels for owner to filter out.